### PR TITLE
Extension on IDataReader<IMultiStreamSource>, and DataReader<IMultiStreamSource, TShape> to read from one or several file paths, rather than requiring constructing an IMultiStreamSource

### PR DESCRIPTION
**Pull request by sfilipi**
Addresses part of  #1090  in this first iteration, by adding a Read extension method taking a string as param, on the IDataReader<IMultiStreamSource>. 

Still looking at how to make so all the derived classes of DataReader<IMultiStreamSource, TShape> can have a Read that takes a string as param.

### CODE REVIEW

**Comment by sfilipi** [comment]
@TomFinley looking at the second part of your suggestion on #1041, on creating an extension on IDataReaderEstimator<IMultiStreamSource, ...>, would that be only for the Fit method?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs [hunk 1]
```diff
@@ -256,5 +256,12 @@ public Column Create()
             }
         }
     }
-}
 
+    public static class LocalPathReader
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs]: >LocalPathReader [](start = 24, length = 15)

I don't think it belongs to `TextLoaderStatic`.

**Comment by Zruty0** [comment]
What do you think about updating the cookbook and associated test cases?

---
In reply to: [430700966](https://github.com/dotnet/machinelearning/pull/1281#issuecomment-430700966) [](ancestors = 430700966)

**Comment by sfilipi** [comment]
The examples in the cookbook use [Read in the DataReader type](https://github.com/dotnet/machinelearning/blob/ed54adeb790cd9385a2ab73b095d0eb229051a91/src/Microsoft.ML.Data/StaticPipe/DataReader.cs#L44) . 
Not sure how to go about that, see the question to Tom above. 

Can't make an extension for that because of the TShape generic. 

---
In reply to: [430768483](https://github.com/dotnet/machinelearning/pull/1281#issuecomment-430768483) [](ancestors = 430768483,430700966)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs (hunk 1 above)
Comment (sfilipi) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs]: so... where does it belong :)

---
In reply to: [226074992](https://github.com/dotnet/machinelearning/pull/1281#discussion_r226074992) [](ancestors = 226074992)

Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs]: Actually, I just added an instance method to `TextLoader` :)

---
In reply to: [226082986](https://github.com/dotnet/machinelearning/pull/1281#discussion_r226082986) [](ancestors = 226082986,226074992)

Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs]: See my MLContext PR

---
In reply to: [226084273](https://github.com/dotnet/machinelearning/pull/1281#discussion_r226084273) [](ancestors = 226084273,226082986,226074992)

**Comment by Zruty0** [comment]
Nevermind than. 

---
In reply to: [430771669](https://github.com/dotnet/machinelearning/pull/1281#issuecomment-430771669) [](ancestors = 430771669,430768483,430700966)

**Comment by sfilipi** [comment]
Nevermind == close this PR?

---
In reply to: [430778815](https://github.com/dotnet/machinelearning/pull/1281#issuecomment-430778815) [](ancestors = 430778815,430771669,430768483,430700966)

**Comment by markusweimer** [comment]
Minor nit: The wording of the PR title might be confusing. When I first saw it, I thought that we allow reading of data from a `string`, while the PR adds an overload to `IDataReader` to accept a filename.

**Comment by eerhardt** [comment]
>     /// multiple paths separated by +.

Does this comment still apply?

---
Refers to: src/Microsoft.ML.Data/DataLoadSave/MultiFileSource.cs:13 in 0425636. [](commit_id = 0425636260d23cb12ca2c3732849f48efdbfef52, deletion_comment = False)

File: src/Microsoft.ML.Data/DataLoadSave/MultiFileSource.cs [hunk 2]
```diff
@@ -16,19 +16,29 @@ public sealed class MultiFileSource : IMultiStreamSource
...
-            if (_paths.Length == 0)
-                throw Contracts.ExceptIO("Could not find file '{0}'", path);
+            // in case of usage from Maml, the paths would be wildcard concatenated in the
+            // first string of paths.
+            string[] concatenated = paths[0] != null ? StreamUtils.ExpandWildCards(paths[0]) : null;
+
+            if (concatenated != null && concatenated.Length > 1)
+            {
+                if (paths.Length > 1)
+                    throw Contracts.Except($"Pass a single string to the {nameof(MultiFileSource)} constructor, if you are using wildcards.");
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/DataLoadSave/MultiFileSource.cs]: It would be nice to put this in xml doc comments.

**Comment by sfilipi** [comment]
>     /// multiple paths separated by +.

yes, due to MAML backwards compat we need to support multiple paths. 

---
In reply to: [431087633](https://github.com/dotnet/machinelearning/pull/1281#issuecomment-431087633) [](ancestors = 431087633)

---
Refers to: src/Microsoft.ML.Data/DataLoadSave/MultiFileSource.cs:13 in 0425636. [](commit_id = 0425636260d23cb12ca2c3732849f48efdbfef52, deletion_comment = False)

**Pull request merged**